### PR TITLE
Replace SummitDev recipe with Summit

### PR DESCRIPTION
**Pull request by prckent**
The existing SummitDev recipe can not work on Summit since the XL compiler is not C++14 capable.

### CODE REVIEW

File: manual/installation.tex [hunk 1]
```diff
@@ -800,33 +800,32 @@ \subsection{Installing on ORNL OLCF Eos Cray XC30}
...
 export FFTW_HOME=$OLCF_FFTW_ROOT
-module load python
 module load cmake
 module load boost
 module load cuda
-mkdir build_summitdev
-cd build_summitdev
+mkdir build_summit
+cd build_summit
 cmake -DCMAKE_C_COMPILER="mpixlc" \
```
Comment (atillack) [inline manual/installation.tex]: This should be mpicc.

File: manual/installation.tex [hunk 2]
```diff
@@ -800,33 +800,32 @@ \subsection{Installing on ORNL OLCF Eos Cray XC30}
...
       -DCMAKE_CXX_COMPILER="mpixlC" \
```
Comment (atillack) [inline manual/installation.tex]: mlicxx

**Comment by atillack** [comment]
@prckent It successfully built. The last time I built using a recipe similar to this I ended up with lots of NaNs which went away when not using netlib-lapack. I hope this is fixed now but since the machine is blocked until tomorrow morning I can't run.

**Comment by trudeaun** [comment]
Which C++14 language features do you need support for? Unfortunately, as you noticed, IBM XL C/C++ for Linux 16.1.1 only offers [partial support for C++14](https://www.ibm.com/support/knowledgecenter/en/SSXVZZ_16.1.1/com.ibm.xlcpp1611.lelinux.doc/language_ref/standard_features.html).

**Comment by ye-luo** [comment]
Does this pass the unit tests on the login node?

**Comment by prckent** [comment]
@ye-luo Yes
```
build_summit_gpu]$ date; nice ctest -R unit
Mon Dec 17 15:01:14 EST 2018
Test project ...  /git_QMCPACK_prckent/build_summit_gpu
      Start  1: deterministic-unit_test_ohmmspete
 1/18 Test  #1: deterministic-unit_test_ohmmspete .......   Passed    0.24 sec
      Start  2: deterministic-unit_test_numerics
 2/18 Test  #2: deterministic-unit_test_numerics ........   Passed    0.21 sec
      Start  3: deterministic-unit_test_utilities
 3/18 Test  #3: deterministic-unit_test_utilities .......   Passed    0.22 sec
      Start  4: deterministic-unit_test_einspline
 4/18 Test  #4: deterministic-unit_test_einspline .......   Passed    2.07 sec
      Start  5: deterministic-unit_test_spline2
 5/18 Test  #5: deterministic-unit_test_spline2 .........   Passed    0.22 sec
      Start  6: deterministic-unit_test_ohmmsdata
 6/18 Test  #6: deterministic-unit_test_ohmmsdata .......   Passed    0.23 sec
      Start  7: deterministic-unit_test_io
 7/18 Test  #7: deterministic-unit_test_io ..............   Passed    0.55 sec
      Start  8: deterministic-unit_test_lattice
 8/18 Test  #8: deterministic-unit_test_lattice .........   Passed    0.22 sec
      Start  9: deterministic-unit_test_particle
 9/18 Test  #9: deterministic-unit_test_particle ........   Passed    2.35 sec
      Start 10: deterministic-unit_test_particleio
10/18 Test #10: deterministic-unit_test_particleio ......   Passed    0.48 sec
      Start 11: deterministic-unit_test_wavefunction
11/18 Test #11: deterministic-unit_test_wavefunction ....   Passed    2.39 sec
      Start 12: deterministic-unit_test_hamiltonian
12/18 Test #12: deterministic-unit_test_hamiltonian .....   Passed   14.73 sec
      Start 13: deterministic-unit_test_type_traits
13/18 Test #13: deterministic-unit_test_type_traits .....   Passed    0.22 sec
      Start 14: deterministic-unit_test_particle_base
14/18 Test #14: deterministic-unit_test_particle_base ...   Passed    0.47 sec
      Start 15: deterministic-unit_test_estimators
15/18 Test #15: deterministic-unit_test_estimators ......   Passed    2.33 sec
      Start 16: deterministic-unit_test_drivers
16/18 Test #16: deterministic-unit_test_drivers .........   Passed    2.21 sec
      Start 17: deterministic-unit_test_qmcapp
17/18 Test #17: deterministic-unit_test_qmcapp ..........   Passed    2.32 sec
      Start 18: deterministic-unit_test_message
18/18 Test #18: deterministic-unit_test_message .........   Passed    0.48 sec

100% tests passed, 0 tests failed out of 18

Label Time Summary:
unit    =  31.95 sec*proc (18 tests)

Total Test time (real) =  32.08 sec
```

File: manual/installation.tex (hunk 1 above)
Comment (prckent) [reply manual/installation.tex]: Updated. However, note that that programming environment happily does the right thing without the change.

File: manual/installation.tex (hunk 2 above)
Comment (prckent) [reply manual/installation.tex]: Updated. However, note that that programming environment happily does the right thing without the change.

**Comment by prckent** [comment]
@trudeaun Will find out. Not many C++14 features yet.

We also have a hard check on the existence of a C++14 standard library during configure, by checking whether access tuple by type is supported https://github.com/QMCPACK/qmcpack/pull/1135 . I haven't tried pointing the compiler towards the GNU implementation yet - is this the preferred solution currently?

**Comment by markdewing** [comment]
@trudeaun The first step that doesn't work is specifying c++14 from the command line with CMake.  (CMake uses the documented flag, the compiler rejects it).   Since that doesn't work, it's hard to find out what other C++14 features do or do not work.

**Comment by prckent** [comment]
Since Summit is down until Thursday and we need to make the release, please can we merge this.

**Comment by trudeaun** [comment]
@markdewing Can you confirm that using `set(CMAKE_CXX_STANDARD 14)` with CMake is not properly invoking XL compiler option `-std=c++1y`? Which XL compiler version are you using? I can open up an issue with CMake if it's not being passed properly, and in the meantime, you might try just using [`-std=c++1y`](https://www.ibm.com/support/knowledgecenter/SSXVZZ_16.1.1/com.ibm.xlcpp1611.lelinux.doc/compiler_ref/opt_langlvl.html) directly if possible. 

If you get any compiler diagnostics indicating that unsupported C++14 features were found in your code, I would appreciate it if you could let me know which ones they are so we can prioritize them in XL's plan for future C++14 support. You will have to switch to another compiler with the needed support (such as gcc) to get the support until then. Thanks!

**Comment by markdewing** [comment]
@trudeaun Here's the text of what I sent to OLCF help a few weeks ago.

Version detail:
IBM XL C/C++ for Linux, V16.1.1 (Beta 6)
Version: 16.01.0001.0000

The man page says 'c++1y' should enable C++ 14 support, but the command line does not recognize it.
[mdewing@login4.summit c14test]$ xlC -std=c++1y tmp.cpp 
warning: 1540-5203 Unrecognized value "c++1y" specified with option "std".

Using 'extended1y' works, but I'm not sure how to get CMake to use that option instead of "c++1y"
Finally, when using the working flag, the compiler picks up system g++ include files, which do not support C++14.   Most compilers have a way to specify a different gcc install to get the include files, but I didn't see anything in the man page.   Or on some machines, loading a newer gcc module would work, but that doesn't work on Summit because loading gcc unloads the XL compiler.

For example, this program
```
#include <type_traits>
#include <vector>
#include <tuple>


int main()
{
    // C++11 feature
    std::vector<int> a = {1,2,3};
    std::tuple<int, double> b(3,4.5);
    // C++14 feature
    int i = std::get<int>(b);
    return 0;
}
```

Compiled with C++14 gives the following errors
```
[mdewing@login4.summit c14test]$ xlC -std=extended1y tmp.cpp 
tmp.cpp:13:13: error: no matching function for call to 'get'
    int i = std::get<int>(b);
            ^~~~~~~~~~~~~
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/utility:142:5: note: candidate
      template ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(std::pair<_Tp1, _Tp2>& __in) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/utility:147:5: note: candidate
      template ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(std::pair<_Tp1, _Tp2>&& __in) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/utility:152:5: note: candidate
      template ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(const std::pair<_Tp1, _Tp2>& __in) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/array:268:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(array<_Tp, _Nm>& __arr) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/array:277:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(array<_Tp, _Nm>&& __arr) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/array:285:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '_Int'
    get(const array<_Tp, _Nm>& __arr) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/tuple:759:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '__i'
    get(tuple<_Elements...>& __t) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/tuple:766:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '__i'
    get(const tuple<_Elements...>& __t) noexcept
    ^
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/tuple:773:5: note: candidate template
      ignored: invalid explicitly-specified argument for template parameter '__i'
    get(tuple<_Elements...>&& __t) noexcept
    ^
1 error generated.
Error while processing tmp.cpp.
```

**Comment by rzurob** [comment]
There was a known issue in Beta 6 that disabled the options for enabling C++14 subset support.  That was corrected in Beta 7 and above.  The -std=c++1y, -std=c++14, -std=gnu++1y, and -std=gnu++14 all work in Beta 7 and above.

Note:  The beta program for 16.1.1 has ended and the betas will expire at the end of the year.  You can download the free-of-charge 16.1.1 community edition from here:  https://www.ibm.com/developerworks/downloads/r/xlcpluslinux/index.html
The community edition is full-featured, but comes without IBM service and support.  The fully-licensed edition adds service and support.

**Comment by prckent** [comment]
Thanks @rzurob . This is helpful. When we get back on Summit@OLCF after the latest software drop we'll reevaluate the situation.

**Comment by rzurob** [comment]
Do you mean you have a newer gcc installation in a non-default directory?  If so, instead of using -`qnostdinc++ -I`, try reconfiguring the compiler by running `xlc_configure` and pointing to the new gcc installation.  The `xlc_configure` tool is documented here:  https://www.ibm.com/support/knowledgecenter/SSXVZZ_16.1.1/com.ibm.xlcpp1611.lelinux.doc/install/configure_options.html

To find out how your compiler was configured:
1. Find out which compiler configuration file you normally use by:
xlc -qversion -V
The XL_CONFIG should show you the absolute path of the configuration file.  In the command above, replace xlc with the invocation command you normally use.

2. Find out the command that was used to generate the configuration file
At the top of the configuration file, there is a comment showing the command that was used to generate it.

3. Generate a new configuration file using the command from step 2 but with -gcc pointing to your new gcc, and -o pointing to an output file name you have write access to (e.g. -o $HOME/xlC.cfg)

4. Test the new cfg file by using the -F option.  (e.g. xlC -F $HOME/xlC.cfg file.cpp)

**Comment by rzurob** [comment]
I just heard back from our C/C++ compiler team.  There might be a simpler way:  Try using the -qgcc_cpp_stdinc and -qgcc_c_stdinc options (which are also called -isystem) instead of -nostdinc++.

https://www.ibm.com/support/knowledgecenter/SSXVZZ_16.1.1/com.ibm.xlcpp1611.lelinux.doc/compiler_ref/opt_gcc_c_stdinc.html

https://www.ibm.com/support/knowledgecenter/SSXVZZ_16.1.1/com.ibm.xlcpp1611.lelinux.doc/compiler_ref/opt_gcc_cpp_stdinc.html

**Comment by markdewing** [comment]
@rzurob Both methods worked for pointing to newer GCC installation. 

The next issue is the __cplusplus variable seemed set at 201103 even if the C++ 14 standard was specified on the command line.  Setting it manually (`-D__cplusplus=201402`) is a workaround.

**Comment by rzurob** [comment]
Yep, that's intentional.  It's because XL C/C++ for Linux V16.1.1 does not fully support C++14 yet.  So we kept the macro at the lower level.  Your workaround is correct.  We also added a -qxflag option to set the level starting in Beta 8:
> - Even under the experimental C++14 language levels, the __cplusplus macro defaults to 201103L. To use the C++14 value of this macro, which may be necessary to access C++14 library interfaces, specify -qxflag=disable__cplusplusOverride or -D__cplusplus=201402L.

While we don't fully support C++14 on Linux yet, we're close.  The Linux release came out in November.  The 16.1 for AIX release which came out a few days ago is C++14 compliant.

**Pull request merged**